Add comments in participatory space presentation page stats block

roxanaopr opened the pull request:
#### :tophat: What? Why?
Add comments count for each topic on the About page, similar to the homepage statistics block.

#### :pushpin: Meta Proposal
https://meta.decidim.org/processes/roadmap/f/122/proposals/16344

#### :pushpin: Related Issues
*Link your PR to an issue*
- Related to #?
- Fixes #?

#### Testing
*Describe the best way to test or validate your PR.*

#### :clipboard: Checklist
:rotating_light: Please review the [guidelines for contributing](https://github.com/decidim/decidim/blob/develop/CONTRIBUTING.adoc) to this repository.

- [ ] :question: **CONSIDER** adding a unit test if your PR resolves an issue.
- [ ] :heavy_check_mark: **DO** check open PR's to avoid duplicates.
- [ ] :heavy_check_mark: **DO** keep pull requests small so they can be easily reviewed.
- [ ] :heavy_check_mark: **DO** build locally before pushing.
- [ ] :heavy_check_mark: **DO** make sure tests pass.
- [ ] :heavy_check_mark: **DO** make sure any new changes are documented in `docs/`.
- [ ] :heavy_check_mark: **DO** add and modify seeds if necessary.
- [ ] :heavy_check_mark: **DO** add CHANGELOG upgrade notes if required.
- [ ] :heavy_check_mark: **DO** add to GraphQL API if there are new public fields.
- [ ] :heavy_check_mark: **DO** add link to MetaDecidim if it's a new feature.
- [ ] :x:**AVOID** breaking the continuous integration build.
- [ ] :x:**AVOID** making significant changes to the overall architecture.

### :camera: Screenshots
![Screenshot from 2021-05-26 14-27-53](https://user-images.githubusercontent.com/66411127/119652488-d5d1c680-be2e-11eb-9f21-86632f230dfc.png)


:hearts: Thank you!

leio10 reviewed (dismissed):
Nice implementation! 👏 

I like the final approach, that doesn't add dependencies for any specific module in core. But I have some doubts before merging it:
* The title and the related proposal refer to "participatory spaces", but you are only adding the stats to "participatory processes". Should we add it also to assemblies and other spaces?
* I saw that `proposals` and `budgets` already have the `comments_count` stats registered, so I understand that you are adding the missing ones (`meetings`). But I also found that `accountability` and `blogs` also have the `comments_count` column. Should we register the stats for those too?
* In the screenshot I see two "COMMENTS" stats. That's a bug? It was fixed after taking the screenshot or it's still there?

roxanaopr commented:
> * In the screenshot I see two "COMMENTS" stats. That's a bug? It was fixed after taking the screenshot or it's still there?

I've added the wrong screenshot, sorry. I've added the correct one :D In stats should be only one "Comments" stat

roxanaopr commented:
> * The title and the related proposal refer to "participatory spaces", but you are only adding the stats to "participatory processes". Should we add it also to assemblies and other spaces?
My assumption was that we should add that stat for participatory processes, but I think that we can also add it to other spaces. It will be nice to have all those spaces specified in the meta proposal, because it wasn't clear for me.

carolromero commented:
Hi there! Sorry for the delay @roxanaopr and @leio10.

Regarding your doubts:
> The title and the related proposal refer to "participatory spaces", but you are only adding the stats to "participatory processes". Should we add it also to assemblies and other spaces?

in the proposal it says participatory spaces, so we assume that it refers to all spaces that have statistics, that is: Processes, Assemblies and Conferences.

> I saw that proposals and budgets already have the comments_count stats registered, so I understand that you are adding the missing ones (meetings). But I also found that accountability and blogs also have the comments_count column. Should we register the stats for those too?

Yep, we should count all comments from all components that make use of them.

@roxanaopr do you think it is possible to add it, in this or another PR? Thank you!

roxanaopr commented:
Hello @carolromero! Thank you for the response! I will add in this PR the remaining implementation.

alecslupu commented:
@carolromero  I guess it would be great if you could raise a meta proposal and make sure that all the metrics / stats are aligned between all the participatory spaces.

leio10 commented on decidim-conferences/app/presenters/decidim/conferences/conference_stats_presenter.rb:
```diff
@@ -26,7 +26,7 @@ def component_stats(conditions)
                             .filter(conditions)
                             .with_context(published_components)
-                            .map { |name, data| [name, data] }.flatten
+                            .map { |name, data| [component_manifest.name, name, data] }.flatten
```
💬 I know that you didn't change this, but I don't understand the `.flatten` part. If there are more than one record it will mix all results in just one list? That would return a list of more than three values, but the line 17 expects a list of lists with 3 values. I guess that this is usually working because the query is returning only one result, so the flatten converts `[["manifest_name", "name", "data"]]` into `["manifest_name", "name", "data"]`.

I think it would be better to call `.flatten(1)` in the outer loop (line 30) to be sure that que get a list of lists. I did a test of what I would expect to happen:

```ruby
irb(main):001:0> [[[1,2,3], [1,2,3]], [[2,3,4],[4,5,6]]].flatten(1)
=> [[1, 2, 3], [1, 2, 3], [2, 3, 4], [4, 5, 6]] 
```

What do you think?

roxanaopr replied:
without flatten

`[[], [], [[:proposals, :proposals_count, 7], [:proposals, :proposals_accepted, 1], [:proposals, :supports_count, 5]], [], [], [[:accountability, :results_count, 36]], [[:debates, :debates_count, 6]], [[:sortitions, :sortitions_count, 0]], [], [[:elections, :elections_count, 6]], [], [], [[:meetings, :meetings_count, 8]], [[:proposals, :endorsements_count, 14]], [], [[:surveys, :answers_count, 0]], [], [[:debates, :endorsements_count, 0]], [], [[:blogs, :posts_count, 5]], [], []]`


with flatten

`[[], [], [:proposals, :proposals_count, 7, :proposals, :proposals_accepted, 1, :proposals, :supports_count, 5], [], [], [:accountability, :results_count, 36], [:debates, :debates_count, 6], [:sortitions, :sortitions_count, 0], [], [:elections, :elections_count, 6], [], [], [:meetings, :meetings_count, 8], [:proposals, :endorsements_count, 14], [], [:surveys, :answers_count, 0], [], [:debates, :endorsements_count, 0], [], [:blogs, :posts_count, 5], [], []]`

with flatten(1)

`[[], [], [:proposals, :proposals_count, 7, :proposals, :proposals_accepted, 1, :proposals, :supports_count, 5], [], [], [:accountability, :results_count, 36], [:debates, :debates_count, 6], [:sortitions, :sortitions_count, 0], [], [:elections, :elections_count, 6], [], [], [:meetings, :meetings_count, 8], [:proposals, :endorsements_count, 14], [], [:surveys, :answers_count, 0], [], [:debates, :endorsements_count, 0], [], [:blogs, :posts_count, 5], [], []]`

Seems that for our case, both versions will return the same output. @leio10 I think that we can leave it as it is now, what do you think?

leio10 replied:
Mmm, but that doesn't mean that it's not working? I think that the rest of the code expects tuples of 3 values. I'd expect this method to return:
```ruby
[[:proposals, :proposals_count, 7], [:proposals, :proposals_accepted, 1], [:proposals, :supports_count, 5], [:accountability, :results_count, 36], [:debates, :debates_count, 6], [:sortitions, :sortitions_count, 0], [:elections, :elections_count, 6], [:meetings, :meetings_count, 8], [:proposals, :endorsements_count, 14], [:surveys, :answers_count, 0], [:debates, :endorsements_count, 0], [:blogs, :posts_count, 5]]
```

If that's true, we can just do `.sum` instead of `.flatten(1)`. What do you think?

roxanaopr replied:
In `def statistics(grouped_stats)` is generated the "final" version of the statistics. There, for each manifest name, we extract each statistic and his value. Bellow is the first iteration, for manifest_name: "proposal"

```
backend-app        | :proposals
backend-app        | [[:proposals, :proposals_count, 7, :proposals, :proposals_accepted, 1, :proposals, :supports_count, 5], [:proposals, :endorsements_count, 14]]

backend-app        | proposals
backend-app        | proposals_count
backend-app        | 7
backend-app        | proposals
backend-app        | proposals_accepted
backend-app        | 1
backend-app        | proposals
backend-app        | supports_count
backend-app        | 5
```

So, having all statistics for proposals (or other manifest) on the same list, it's ok. 

Version with "sum" will generate something like this:
```
[0,
 0,
 [:proposals,
  :proposals_count,
  7,
  :proposals,
  :proposals_accepted,
  1,
  :proposals,
  :supports_count,
  5],
 0,
 0,
 [:accountability, :results_count, 36],
 [:debates, :debates_count, 6],
 [:sortitions, :sortitions_count, 0],
 0,
 [:elections, :elections_count, 6],
 0,
 0,
 [:meetings, :meetings_count, 8],
 [:proposals, :endorsements_count, 14],
 0,
 [:surveys, :answers_count, 0],
 0,
 [:debates, :endorsements_count, 0],
 0,
 [:blogs, :posts_count, 5],
 0,
 0]
```
And for sum we will also need to change `highlighted_stats = highlighted_stats.reject { |_manifest, _name, data| data.zero? }`

leio10 replied:
ok, sorry, let's merge it as it is 👍

leio10 commented on decidim-assemblies/app/presenters/decidim/assemblies/assembly_stats_presenter.rb:
```diff
@@ -13,22 +13,12 @@ def collection
         highlighted_stats.concat(assembly_followers_stats(priority: StatsRegistry::HIGH_PRIORITY))
         highlighted_stats.concat(component_stats(priority: StatsRegistry::HIGH_PRIORITY))
         highlighted_stats.concat(component_stats(priority: StatsRegistry::MEDIUM_PRIORITY))
+        highlighted_stats.concat(comments_stats(:assemblies, tag: :comments))
```
💬 I have a doubt: why are you adding the `tag: :comments` condition if you are only retrieving the `:comments_count` stat? It isn't redundant?

leio10 reviewed (dismissed):
LGTM, please let me know your opinion about my doubts.
